env/fs_on_demand.cc

```diff
@@ -0,0 +1,344 @@
+
+    IOStatus s = target()->FileExists(rname, file_opts.io_options, dbg);
+    if (s.ok()) {
+      return IOStatus::InvalidArgument(
+          "Writing to a file present in the remote directory not supoprted");
+    }
+  }
+
+  return target()->NewWritableFile(fname, file_opts, result, dbg);
+}
+
+// Currently not supported, as there's no need for RocksDB to create a
+// directory object for a DB in read-only mode.
+IOStatus OnDemandFileSystem::NewDirectory(
+    const std::string& /*name*/, const IOOptions& /*io_opts*/,
+    std::unique_ptr<FSDirectory>* /*result*/, IODebugContext* /*dbg*/) {
+  return IOStatus::NotSupported();
+}
+
+// Check if the given file exists, either locally or remote. If the file is an
+// SST file, then link it locally. We assume if the file existence is being
+// checked, its for verification purposes, for example while replaying the
+// MANIFEST. The file will be opened for reading  some time in the future.
+IOStatus OnDemandFileSystem::FileExists(const std::string& fname,
+                                        const IOOptions& options,
+                                        IODebugContext* dbg) {
+  IOStatus s = target()->FileExists(fname, options, dbg);
+  if (!s.IsNotFound() && !s.IsPathNotFound()) {
+    return s;
+  }
+
+  std::string rname = fname;
+  if (CheckPathAndAdjust(rname)) {
+    // First clear any local directory cache as it may be out of date
+    target()->DiscardCacheForDirectory(rname);
+
+    FileType type;
+    if (LookupFileType(fname, &type) && type == kTableFile) {
+      s = target()->LinkFile(rname, fname, options, dbg);
+    } else {
+      s = target()->FileExists(rname, options, dbg);
+    }
+  }
+  return s;
+}
+
+// Doa listing of both the local and remote directories and merge the two.
+IOStatus OnDemandFileSystem::GetChildren(const std::string& dir,
+                                         const IOOptions& options,
+                                         std::vector<std::string>* result,
+                                         IODebugContext* dbg) {
+  std::string rdir = dir;
+  IOStatus s = target()->GetChildren(dir, options, result, dbg);
+  if (!s.ok() || !CheckPathAndAdjust(rdir)) {
+    return s;
+  }
+
+  std::vector<std::string> rchildren;
+  // First clear any local directory cache as it may be out of date
+  target()->DiscardCacheForDirectory(rdir);
+  s = target()->GetChildren(rdir, options, &rchildren, dbg);
+  if (s.ok()) {
+    std::for_each(rchildren.begin(), rchildren.end(), [&](std::string& name) {
+      // Adjust name
+      size_t pos = name.find(dest_path_);
+      if (pos == 0) {
+        name.replace(pos, dest_path_.length(), src_path_);
+      }
+    });
+    std::sort(result->begin(), result->end());
+    std::sort(rchildren.begin(), rchildren.end());
+
+    std::vector<std::string> output(result->size() + rchildren.size());
+    std::set_union(result->begin(), result->end(), rchildren.begin(),
+                   rchildren.end(), output.begin());
```

Comment:
I think we might end up with a bunch of empty elements at the end of the vector (as many as the number of files present in both vectors)